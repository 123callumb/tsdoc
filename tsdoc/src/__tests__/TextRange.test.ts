import { TextRange } from '../index';

function matchSnapshot(textRange: TextRange): void {
  for (let i: number = -1; i <= textRange.end + 1; ++i) {
      // Show the next 10 characters
    const context: string = textRange.buffer.substr(Math.max(i, 0), 10)
      .replace(/\n/g, '[n]').replace(/\r/g, '[r]');

    expect({
      i: i,
      c: textRange.buffer[i],
      context: context,
      location: textRange.getLocation(i)
    }).toMatchSnapshot();
  }
}

test('construction scenarios', () => {
  const buffer: string = '0123456789';
  const textRange: TextRange = TextRange.fromString(buffer);
  expect(textRange.toString()).toEqual(buffer);

  const subRange: TextRange = textRange.getNewRange(3, 6);
  expect(subRange).toMatchSnapshot('subRange');
});

test('location calculation', () => {
  const textRange: TextRange = TextRange.fromString([
    'L1',
    'L2',
    '', // (line 3 is blank)
    'L4',
    'L5+CR\rL5+CRLF\r\nL6+LFCR\n\rL7'
  ].join('\n'));
  matchSnapshot(textRange);
});

test('location calculation empty string', () => {
  const textRange: TextRange = TextRange.fromString('');
  matchSnapshot(textRange);
});

test('location calculation newline string', () => {
  const textRange: TextRange = TextRange.fromString('\n');
  matchSnapshot(textRange);
});
