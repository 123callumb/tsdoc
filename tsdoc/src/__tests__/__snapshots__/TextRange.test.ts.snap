// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`construction scenarios: subRange 1`] = `
TextRange {
  "buffer": "0123456789",
  "end": 6,
  "pos": 3,
}
`;

exports[`location calculation 1`] = `
Object {
  "c": undefined,
  "context": "L1[n]L2[n][n]L4[n]",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`location calculation 2`] = `
Object {
  "c": "L",
  "context": "L1[n]L2[n][n]L4[n]",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`location calculation 3`] = `
Object {
  "c": "1",
  "context": "1[n]L2[n][n]L4[n]L",
  "i": 1,
  "location": Object {
    "column": 2,
    "line": 1,
  },
}
`;

exports[`location calculation 4`] = `
Object {
  "c": "
",
  "context": "[n]L2[n][n]L4[n]L5",
  "i": 2,
  "location": Object {
    "column": 3,
    "line": 1,
  },
}
`;

exports[`location calculation 5`] = `
Object {
  "c": "L",
  "context": "L2[n][n]L4[n]L5+",
  "i": 3,
  "location": Object {
    "column": 1,
    "line": 2,
  },
}
`;

exports[`location calculation 6`] = `
Object {
  "c": "2",
  "context": "2[n][n]L4[n]L5+C",
  "i": 4,
  "location": Object {
    "column": 2,
    "line": 2,
  },
}
`;

exports[`location calculation 7`] = `
Object {
  "c": "
",
  "context": "[n][n]L4[n]L5+CR",
  "i": 5,
  "location": Object {
    "column": 3,
    "line": 2,
  },
}
`;

exports[`location calculation 8`] = `
Object {
  "c": "
",
  "context": "[n]L4[n]L5+CR[r]",
  "i": 6,
  "location": Object {
    "column": 1,
    "line": 3,
  },
}
`;

exports[`location calculation 9`] = `
Object {
  "c": "L",
  "context": "L4[n]L5+CR[r]L",
  "i": 7,
  "location": Object {
    "column": 1,
    "line": 4,
  },
}
`;

exports[`location calculation 10`] = `
Object {
  "c": "4",
  "context": "4[n]L5+CR[r]L5",
  "i": 8,
  "location": Object {
    "column": 2,
    "line": 4,
  },
}
`;

exports[`location calculation 11`] = `
Object {
  "c": "
",
  "context": "[n]L5+CR[r]L5+",
  "i": 9,
  "location": Object {
    "column": 3,
    "line": 4,
  },
}
`;

exports[`location calculation 12`] = `
Object {
  "c": "L",
  "context": "L5+CR[r]L5+C",
  "i": 10,
  "location": Object {
    "column": 1,
    "line": 5,
  },
}
`;

exports[`location calculation 13`] = `
Object {
  "c": "5",
  "context": "5+CR[r]L5+CR",
  "i": 11,
  "location": Object {
    "column": 2,
    "line": 5,
  },
}
`;

exports[`location calculation 14`] = `
Object {
  "c": "+",
  "context": "+CR[r]L5+CRL",
  "i": 12,
  "location": Object {
    "column": 3,
    "line": 5,
  },
}
`;

exports[`location calculation 15`] = `
Object {
  "c": "C",
  "context": "CR[r]L5+CRLF",
  "i": 13,
  "location": Object {
    "column": 4,
    "line": 5,
  },
}
`;

exports[`location calculation 16`] = `
Object {
  "c": "R",
  "context": "R[r]L5+CRLF[r]",
  "i": 14,
  "location": Object {
    "column": 5,
    "line": 5,
  },
}
`;

exports[`location calculation 17`] = `
Object {
  "c": "
",
  "context": "[r]L5+CRLF[r][n]",
  "i": 15,
  "location": Object {
    "column": 5,
    "line": 5,
  },
}
`;

exports[`location calculation 18`] = `
Object {
  "c": "L",
  "context": "L5+CRLF[r][n]L",
  "i": 16,
  "location": Object {
    "column": 6,
    "line": 5,
  },
}
`;

exports[`location calculation 19`] = `
Object {
  "c": "5",
  "context": "5+CRLF[r][n]L6",
  "i": 17,
  "location": Object {
    "column": 7,
    "line": 5,
  },
}
`;

exports[`location calculation 20`] = `
Object {
  "c": "+",
  "context": "+CRLF[r][n]L6+",
  "i": 18,
  "location": Object {
    "column": 8,
    "line": 5,
  },
}
`;

exports[`location calculation 21`] = `
Object {
  "c": "C",
  "context": "CRLF[r][n]L6+L",
  "i": 19,
  "location": Object {
    "column": 9,
    "line": 5,
  },
}
`;

exports[`location calculation 22`] = `
Object {
  "c": "R",
  "context": "RLF[r][n]L6+LF",
  "i": 20,
  "location": Object {
    "column": 10,
    "line": 5,
  },
}
`;

exports[`location calculation 23`] = `
Object {
  "c": "L",
  "context": "LF[r][n]L6+LFC",
  "i": 21,
  "location": Object {
    "column": 11,
    "line": 5,
  },
}
`;

exports[`location calculation 24`] = `
Object {
  "c": "F",
  "context": "F[r][n]L6+LFCR",
  "i": 22,
  "location": Object {
    "column": 12,
    "line": 5,
  },
}
`;

exports[`location calculation 25`] = `
Object {
  "c": "
",
  "context": "[r][n]L6+LFCR[n]",
  "i": 23,
  "location": Object {
    "column": 12,
    "line": 5,
  },
}
`;

exports[`location calculation 26`] = `
Object {
  "c": "
",
  "context": "[n]L6+LFCR[n][r]",
  "i": 24,
  "location": Object {
    "column": 13,
    "line": 5,
  },
}
`;

exports[`location calculation 27`] = `
Object {
  "c": "L",
  "context": "L6+LFCR[n][r]L",
  "i": 25,
  "location": Object {
    "column": 1,
    "line": 6,
  },
}
`;

exports[`location calculation 28`] = `
Object {
  "c": "6",
  "context": "6+LFCR[n][r]L7",
  "i": 26,
  "location": Object {
    "column": 2,
    "line": 6,
  },
}
`;

exports[`location calculation 29`] = `
Object {
  "c": "+",
  "context": "+LFCR[n][r]L7",
  "i": 27,
  "location": Object {
    "column": 3,
    "line": 6,
  },
}
`;

exports[`location calculation 30`] = `
Object {
  "c": "L",
  "context": "LFCR[n][r]L7",
  "i": 28,
  "location": Object {
    "column": 4,
    "line": 6,
  },
}
`;

exports[`location calculation 31`] = `
Object {
  "c": "F",
  "context": "FCR[n][r]L7",
  "i": 29,
  "location": Object {
    "column": 5,
    "line": 6,
  },
}
`;

exports[`location calculation 32`] = `
Object {
  "c": "C",
  "context": "CR[n][r]L7",
  "i": 30,
  "location": Object {
    "column": 6,
    "line": 6,
  },
}
`;

exports[`location calculation 33`] = `
Object {
  "c": "R",
  "context": "R[n][r]L7",
  "i": 31,
  "location": Object {
    "column": 7,
    "line": 6,
  },
}
`;

exports[`location calculation 34`] = `
Object {
  "c": "
",
  "context": "[n][r]L7",
  "i": 32,
  "location": Object {
    "column": 8,
    "line": 6,
  },
}
`;

exports[`location calculation 35`] = `
Object {
  "c": "
",
  "context": "[r]L7",
  "i": 33,
  "location": Object {
    "column": 8,
    "line": 6,
  },
}
`;

exports[`location calculation 36`] = `
Object {
  "c": "L",
  "context": "L7",
  "i": 34,
  "location": Object {
    "column": 1,
    "line": 7,
  },
}
`;

exports[`location calculation 37`] = `
Object {
  "c": "7",
  "context": "7",
  "i": 35,
  "location": Object {
    "column": 2,
    "line": 7,
  },
}
`;

exports[`location calculation 38`] = `
Object {
  "c": undefined,
  "context": "",
  "i": 36,
  "location": Object {
    "column": 3,
    "line": 7,
  },
}
`;

exports[`location calculation 39`] = `
Object {
  "c": undefined,
  "context": "",
  "i": 37,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`location calculation empty string 1`] = `
Object {
  "c": undefined,
  "context": "",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`location calculation empty string 2`] = `
Object {
  "c": undefined,
  "context": "",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`location calculation empty string 3`] = `
Object {
  "c": undefined,
  "context": "",
  "i": 1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`location calculation newline string 1`] = `
Object {
  "c": undefined,
  "context": "[n]",
  "i": -1,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;

exports[`location calculation newline string 2`] = `
Object {
  "c": "
",
  "context": "[n]",
  "i": 0,
  "location": Object {
    "column": 1,
    "line": 1,
  },
}
`;

exports[`location calculation newline string 3`] = `
Object {
  "c": undefined,
  "context": "",
  "i": 1,
  "location": Object {
    "column": 1,
    "line": 2,
  },
}
`;

exports[`location calculation newline string 4`] = `
Object {
  "c": undefined,
  "context": "",
  "i": 2,
  "location": Object {
    "column": 0,
    "line": 0,
  },
}
`;
